### PR TITLE
pylint pre-commit

### LINKED ISSUE #48737: STYLE: add pylint pre-commit hook
Having pylint in CI would've prevented a bug from entering 1.5.0 https://github.com/pandas-dev/pandas/pull/48736

So let's add pylint as a pre-commit hook

pylint has a _lot_ of options - some duplicative of what's already in `flake8`, others arguably too opinionated, others that we'd like to eventually enable but perhaps not all in one go

So, the task here is:
- add `pylint` to https://github.com/pandas-dev/pandas/blob/main/.pre-commit-config.yaml
- run `pre-commit run pylint --all-files`
- fixup any linting errors which come up. If some linting error would require big changes to lots of files, or refactoring, then it's OK to turn it off for now, see [pylint configuration](https://pylint.pycqa.org/en/latest/user_guide/configuration/index.html)
- once `pre-commit run pylint --all-files` passes, then stage, commit, and push! Feel free to tag me in the PR

### PR DESCRIPTION
**Pull request by YvanCywan**
- [ ] closes #48737
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).

Added pylint to CI. Disabled most messages due to the sheer amount of warnings and changes that would need to be fixed in other PR.

@MarcoGorelli

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -59,6 +59,17 @@ repos:
...
+-   repo: https://github.com/pycqa/pylint
+    rev: v2.15.3
+    hooks:
+    -   id: pylint
+        types: [python]
+        args: 
+            [
+                "-rn",
+                "-d=C,R,W,import-error,no-member,no-name-in-module,unsubscriptable-object,invalid-unary-operand-type,unexpected-keyword-arg,not-an-iterable,unsupported-assignment-operation,unsupported-membership-test,no-value-for-parameter,unpacking-non-sequence,bad-super-call,access-member-before-definition,used-before-assignment,too-many-fucntion-args,abstract-class-instantiated,not-callable,too-many-fucntion-args,invalid-repr-returned"
+            ]
```
Comment (MarcoGorelli) [inline .pre-commit-config.yaml:72]: can these be added to `pyproject.toml`? something like
```
[tool.pylint.messages_control]
disable = [
  "import-error",
  "no-member",
]
```
(and all the others in there)

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -59,6 +59,17 @@ repos:
         - flake8==5.0.4
         - flake8-bugbear==22.7.1
         - pandas-dev-flaker==0.5.0
+    # Added pylint, at the moment it is excluding a lot of messages
```
Comment (MarcoGorelli) [inline .pre-commit-config.yaml:62]: I don't think we need this

File: .pre-commit-config.yaml [hunk 3]
```diff
@@ -59,6 +59,17 @@ repos:
...
+-   repo: https://github.com/pycqa/pylint
+    rev: v2.15.3
+    hooks:
+    -   id: pylint
+        types: [python]
```
Comment (MarcoGorelli) [inline .pre-commit-config.yaml:67]: this is already defined, you don't need it

https://github.com/PyCQA/pylint/blob/ca6648a909a7c583cc6c210bcaa126b4f6fb1f3b/.pre-commit-hooks.yaml#L5

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]
thanks for working on this

I've left a few comments, but there's also still a few lint errors

```
************* Module pandas.tests.frame.methods.test_rename
pandas/tests/frame/methods/test_rename.py:339:12: E1124: Argument 'mapper' passed by position and keyword in method call (redundant-keyword-arg)
************* Module pandas.tests.window.test_groupby
pandas/tests/window/test_groupby.py:1082:8: E0102: function already defined line 1067 (function-redefined)
************* Module pandas.tests.io.pytables.test_append
pandas/tests/io/pytables/test_append.py:440:8: E0102: function already defined line 364 (function-redefined)
************* Module pandas.tests.extension.decimal.test_decimal
pandas/tests/extension/decimal/test_decimal.py:287:4: E0213: Method should have "self" as first argument (no-self-argument)
************* Module doc.make
doc/make.py:87:22: E0602: Undefined variable 'pandas' (undefined-variable)
************* Module pandas.tests.io.test_gcs
pandas/tests/io/test_gcs.py:36:8: E0211: Method has no argument (no-method-argument)
************* Module pandas.tests.groupby.test_groupby
pandas/tests/groupby/test_groupby.py:140:4: E0102: function already defined line 124 (function-redefined)
pandas/tests/groupby/test_groupby.py:172:4: E0102: function already defined line 165 (function-redefined)
pandas/tests/groupby/test_groupby.py:182:4: E0102: function already defined line 165 (function-redefined)
pandas/tests/groupby/test_groupby.py:193:4: E0102: function already defined line 165 (function-redefined)
pandas/tests/groupby/test_groupby.py:204:4: E0102: function already defined line 165 (function-redefined)
************* Module pandas.core.internals.blocks
pandas/core/internals/blocks.py:1984:12: E0704: The raise statement is not inside an except clause (misplaced-bare-raise)
************* Module pandas.tests.io.test_html
pandas/tests/io/test_html.py:1143:8: E1124: Argument 'flavor' passed by position and keyword in function call (redundant-keyword-arg)
pandas/tests/io/test_html.py:1281:12: E0301: __iter__ returns non-iterator (non-iterator-returned)
************* Module pandas.io.clipboard
pandas/io/clipboard/__init__.py:131:17: E0602: Undefined variable 'Foundation' (undefined-variable)
pandas/io/clipboard/__init__.py:132:44: E0602: Undefined variable 'Foundation' (undefined-variable)
pandas/io/clipboard/__init__.py:133:16: E0602: Undefined variable 'AppKit' (undefined-variable)
pandas/io/clipboard/__init__.py:134:35: E0602: Undefined variable 'AppKit' (undefined-variable)
pandas/io/clipboard/__init__.py:135:40: E0602: Undefined variable 'AppKit' (undefined-variable)
pandas/io/clipboard/__init__.py:139:16: E0602: Undefined variable 'AppKit' (undefined-variable)
pandas/io/clipboard/__init__.py:140:39: E0602: Undefined variable 'AppKit' (undefined-variable)
************* Module pandas.tests.scalar.period.test_period
pandas/tests/scalar/period/test_period.py:843:4: E0213: Method should have "self" as first argument (no-self-argument)
************* Module pandas.tests.resample.test_resampler_grouper
pandas/tests/resample/test_resampler_grouper.py:83:4: E0102: function already defined line 68 (function-redefined)
pandas/tests/resample/test_resampler_grouper.py:266:4: E0102: function already defined line 260 (function-redefined)
************* Module pandas.core.arrays.base
pandas/core/arrays/base.py:96:4: E0239: Inheriting "'ExtensionArray'", which is not a class. (inherit-non-class)
************* Module pandas.tests.arrays.datetimes.test_constructors
pandas/tests/arrays/datetimes/test_constructors.py:23:26: E1121: Too many positional arguments for method call (too-many-function-args)
************* Module pandas.tests.io.test_sql
pandas/tests/io/test_sql.py:1609:4: E0213: Method should have "self" as first argument (no-self-argument)
************* Module pandas.core.indexes.base
pandas/core/indexes/base.py:1144:23: E1121: Too many positional arguments for method call (too-many-function-args)
pandas/core/indexes/base.py:5126:19: E1121: Too many positional arguments for method call (too-many-function-args)
pandas/core/indexes/base.py:6478:19: E1121: Too many positional arguments for method call (too-many-function-args)
************* Module pandas.tests.arrays.timedeltas.test_constructors
pandas/tests/arrays/timedeltas/test_constructors.py:14:27: E1121: Too many positional arguments for method call (too-many-function-args)
************* Module pandas.tests.groupby.aggregate.test_aggregate
pandas/tests/groupby/aggregate/test_aggregate.py:297:4: E0102: function already defined line 284 (function-redefined)
************* Module pandas.tests.frame.methods.test_assign
pandas/tests/frame/methods/test_assign.py:70:12: E1121: Too many positional arguments for method call (too-many-function-args)

-----------------------------------
Your code has been rated at 9.99/10
```

**Comment by datapythonista** [comment]
I guess we can organize a sprint for beginners if there are many easy things to fix to make pylint happy.

File: .pre-commit-config.yaml (hunk 3 above)
Comment (HebiKotei) [reply .pre-commit-config.yaml]: Removed

File: .pre-commit-config.yaml (hunk 2 above)
Comment (HebiKotei) [reply .pre-commit-config.yaml]: Removed

File: .pre-commit-config.yaml (hunk 1 above)
Comment (HebiKotei) [reply .pre-commit-config.yaml]: Done

**Comment by HebiKotei** [comment]
I have updated the request with the changes you have suggested, I'll start working on the lint errors.

**Comment by HebiKotei** [comment]
Have fixed on raise function, however, pylint passes locally rather than on CI, due to the inline disables I have added in some locations.

**Comment by MarcoGorelli** [comment]
looks like you just need to run the other checks as well, `pre-commit run --all-files`

**Review by MarcoGorelli — APPROVED** [review]
looks good to me, thanks @HebiKotei

**Comment by datapythonista** [comment]
Sorry for the probably silly question, but seem like we're adding pylint, but we aren't fixing any error because of it, only adding marks to disable to validation in some places. Are we planning to fix the checks being disabled in this PR later, or how is this useful?

**Comment by MarcoGorelli** [comment]
Yeah I think they could gradually enabled in sprints - this is at least gives us something start with

**Pull request merged**